File: public/benchmarks/native/20260913T222757Z/harnesses/go/lm15/main.go

package main
import (
    "context"
    "encoding/json"
    "fmt"
    "os"
    "strconv"
    "strings"
    "time"
    lm15 "github.com/lm15-dev/lm15-go"
)
func main() {
    ctx := context.Background()
    endpoint := os.Getenv("LM15_BENCH_ENDPOINT")
    iterations, err := strconv.Atoi(os.Getenv("LM15_BENCH_ITERATIONS")); if err != nil { panic(err) }
    client, err := lm15.NewOpenAIChatLM(lm15.WithAPIKey("benchmark"), lm15.WithBaseURL(endpoint+"/v1")); if err != nil { panic(err) }; defer client.Close()
    call := func() (string, error) { response, err := client.Complete(ctx, &lm15.Request{Model: "gpt-4.1-mini", Messages: []lm15.Message{lm15.UserMessage("Say hello.")}, Config: lm15.Config{MaxTokens: lm15.I(32)}})
        if err != nil { return "", err }
        return response.TextOr(""), nil }
    for i := 0; i < 20; i++ { text, err := call(); if err != nil { panic(err) }; if text != "Hello." { panic("incorrect reply") } }
    started := time.Now()
    for i := 0; i < iterations; i++ { text, err := call(); if err != nil { panic(err) }; if text != "Hello." { panic("incorrect reply") } }
    elapsed := float64(time.Since(started).Nanoseconds()) / 1000 / float64(iterations)
    status, err := os.ReadFile("/proc/self/status"); if err != nil { panic(err) }
    var rss float64; var cpu string
    for _, line := range strings.Split(string(status), "\n") {
        if strings.HasPrefix(line, "VmRSS:") { rss, err = strconv.ParseFloat(strings.Fields(line)[1], 64); if err != nil { panic(err) } }
        if strings.HasPrefix(line, "Cpus_allowed_list:") { cpu = strings.Fields(line)[1] }
    }
    result, err := json.Marshal(map[string]any{"request_us": elapsed, "rss_mib": rss/1024, "cpu_affinity": cpu, "iterations": iterations}); if err != nil { panic(err) }
    fmt.Println("BENCH_RESULT="+string(result))
}

File: public/benchmarks/native/20260913T222757Z/harnesses/rust/lm15/src/main.rs

use std::{env, time::Instant};
use lm15::{OpenAIChatLM, Request, Message};
#[tokio::main(flavor = "current_thread")]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let endpoint = env::var("LM15_BENCH_ENDPOINT")?;
    let iterations: usize = env::var("LM15_BENCH_ITERATIONS")?.parse()?;
    let client = OpenAIChatLM::builder().api_key("benchmark").base_url(format!("{endpoint}/v1")).build()?;
    for _ in 0..20 {
        let text: String = { let request = Request { model: "gpt-4.1-mini".into(), messages: vec![Message::user("Say hello.")?], config: lm15::Config { max_tokens: Some(32), ..Default::default() }, ..Default::default() };
        let response = client.complete(&request).await?;
        response.text().unwrap_or_default().to_string() };
        assert_eq!(text, "Hello.");
        std::hint::black_box(text);
    }
    let started = Instant::now();
    for _ in 0..iterations {
        let text: String = { let request = Request { model: "gpt-4.1-mini".into(), messages: vec![Message::user("Say hello.")?], config: lm15::Config { max_tokens: Some(32), ..Default::default() }, ..Default::default() };
        let response = client.complete(&request).await?;
        response.text().unwrap_or_default().to_string() };
        assert_eq!(text, "Hello.");
        std::hint::black_box(text);
    }
    let request_us = started.elapsed().as_secs_f64() * 1e6 / iterations as f64;
    let status = std::fs::read_to_string("/proc/self/status")?;
    let rss: f64 = status.lines().find(|line| line.starts_with("VmRSS:")).ok_or("RSS")?.split_whitespace().nth(1).ok_or("RSS value")?.parse()?;
    let cpu = status.lines().find(|line| line.starts_with("Cpus_allowed_list:")).ok_or("affinity")?.split_whitespace().nth(1).ok_or("CPU value")?;
    println!("BENCH_RESULT={{\"request_us\":{request_us},\"rss_mib\":{},\"cpu_affinity\":\"{cpu}\",\"iterations\":{iterations}}}", rss / 1024.0);
    Ok(())
}
